source/server/configuration_impl.cc

```diff
@@ -90,6 +90,15 @@ void MainImpl::initialize(const envoy::config::bootstrap::v3::Bootstrap& bootstr
     server.secretManager().addStaticSecret(secrets[i]);
   }
 
+  const auto& certificate_providers = bootstrap.certificate_provider_instances();
+  for (auto iter = certificate_providers.begin(); iter != certificate_providers.end(); iter++ ) {
+    ENVOY_LOG(debug, "===========================================================");
+    ENVOY_LOG(debug, "certificate provider: {}", iter->first);
+    envoy::config::core::v3::TypedExtensionConfig cert_provider_config;
+    cert_provider_config.MergeFrom(iter->second);
```

Comment:
No need to copy?

Reply:
I don't think it needs copy. What's your concern?

Reply:
You could just do
```
server.certificateProviderManager().addCertificateProvider(iter->first, iter->second);
```

instead of creating this copy.